connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java

```diff
@@ -755,4 +777,40 @@ private static Object convertToConnect(Schema schema, JsonNode jsonValue) {
     private interface LogicalTypeConverter {
         Object convert(Schema schema, Object value);
     }
+
+    private void flatten(String key, JsonNode jsonValue, LinkedHashMap<String, Object> map, SchemaBuilder builder) {
```

Comment:
What happens when you have a simple JSON value like string values (e.g., `"foo"`), `null`, `true`, `false`, numbers (e.g., `1234.5`), or even an array? This method should handle all potential JSON values, not just JSON object values.

To do this correctly, we may need to separate the schema inference from the value construction. And if we introduce a method that infers the `Schema` from the JSON value (and this method could be recursive), then we should also be able to reuse the existing `convertToConnect(Schema schema, JsonNode jsonValue)` method. It's not quite as fast since we're iterating over the value twice, but this is actually how the schema-enabled version of the code works and it would reuse a lot more existing code.

WDYT?

Reply:
Thanks for the input, @rhauch!

My focus so far in this initial version of schemaless json support was to enable data ingestion via JDBCSinkConnector without the explicit need for producers to provide schemas. This means only ever resulting in Kafka Connect Structs at the outermost level, which means there's an assumption made that the record is a JSON object. I do agree there is some opportunity to remove code duplication with the existing convertToConnect method. Let me see what I can do to make this more generic for simple JSON values.

Reply:
Thanks for considering my suggestions. Looking forward to the changes!

As for only handling structs, we can't constrain how people use this or what goes into their streams that they consume with a Connect sink using this connector, except to say that this should handle all all possible valid JSON documents, including those comprised of simple JSON values. :-)